File: app/src/main/java/com/elyeproj/webappinteraction/MainActivity.java
package com.elyeproj.webappinteraction;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.webkit.JavascriptInterface;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import java.io.IOException;

public class MainActivity extends AppCompatActivity {

    private static final String JAVASCRIPT_OBJ = "javascript_obj";
    private static final String BASE_URL = "file:///android_asset/webview.html";

    private WebView webView;
    private TextView textFromWeb;
    private Button btnToWeb;
    private EditText edtToWeb;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        webView = (WebView) findViewById(R.id.my_web_view);
        textFromWeb = (TextView) findViewById(R.id.txt_from_web);
        btnToWeb = (Button) findViewById(R.id.btn_send_to_web);
        edtToWeb = (EditText) findViewById(R.id.edit_text_to_web);

        WebSettings settings = webView.getSettings();
        settings.setJavaScriptEnabled(true);

        webView.addJavascriptInterface(new JavaScriptInterface(), JAVASCRIPT_OBJ);

        webView.setWebViewClient(new WebViewClient(){
            @Override
            public void onPageFinished(WebView view, String url) {
                if (url.equals(BASE_URL)) {
                    injectJavaScriptFunction();
                }
            }
        });

        webView.loadUrl(BASE_URL);

        btnToWeb.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                webView.evaluateJavascript("javascript: " +
                        "updateFromAndroid(\"" + edtToWeb.getText() + "\")", null);
            }
        });
    }

    private void injectJavaScriptFunction() {
        webView.loadUrl("javascript: " +
                "window.androidObj.textToAndroid = function(message) { " +
                JAVASCRIPT_OBJ + ".textFromWeb(message) }");
    }


    private class JavaScriptInterface {
        @JavascriptInterface
        public void textFromWeb(String fromWeb) {
            textFromWeb.setText(fromWeb);
        }
    }
}

File: app/src/main/java/com/elyeproj/webappinteraction/MainActivity.kt
package com.elyeproj.webappinteraction

import android.content.pm.ApplicationInfo
import android.os.Build
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.webkit.JavascriptInterface
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import kotlinx.android.synthetic.main.activity_main.*

import java.io.IOException

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            if (0 != (applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE)) {
                WebView.setWebContentsDebuggingEnabled(true)
            }
        }

        my_web_view.settings.javaScriptEnabled = true
        my_web_view.addJavascriptInterface(JavaScriptInterface(), JAVASCRIPT_OBJ)
        my_web_view.webViewClient = object : WebViewClient() {
            override fun onPageFinished(view: WebView, url: String) {
                if (url == BASE_URL) {
                    injectJavaScriptFunction()
                }
            }
        }
        my_web_view.loadUrl(BASE_URL)

        btn_send_to_web.setOnClickListener {
            my_web_view.evaluateJavascript("javascript: " +
                    "updateFromAndroid(\"" + edit_text_to_web.text + "\")", null)
        }
    }

    override fun onDestroy() {
        my_web_view.removeJavascriptInterface(JAVASCRIPT_OBJ)
        super.onDestroy()
    }

    private fun injectJavaScriptFunction() {
        my_web_view.loadUrl("javascript: " +
                "window.androidObj.textToAndroid = function(message) { " +
                JAVASCRIPT_OBJ + ".textFromWeb(message) }")
    }


    private inner class JavaScriptInterface {
        @JavascriptInterface
        fun textFromWeb(fromWeb: String) {
            txt_from_web.text = fromWeb
        }
    }

    companion object {
        private val JAVASCRIPT_OBJ = "javascript_obj"
        private val BASE_URL = "file:///android_asset/webview.html"
    }
}
